### PR TITLE
Demo sandbox milestone 1

### PR DESCRIPTION
**Pull request by TheTallJerry**
First milestone for demo sandbox:

- Set up webpack for hot reloading, installed material UI
- Set up basic UI components with minimal styling and fuctional user flow for automated draw
- Add error boundary to catch rendering errors and disable Webpack `devServer` `Overlay` to show the error boundary's message as top level

### CODE REVIEW

File: demo/src/App.tsx [hunk 1]
```diff
@@ -20,13 +21,22 @@ const DEMO_OBJECTS = [
 ];
 
 export default function App() {
+    const [formData, setFormData] = useState("");
+    const [jsonResult, setJsonResult] = useState([]);
+
+    const onSubmit = async (event) => {
+        event.preventDefault();
+        setJsonResult(JSON.parse(formData));
```
Comment (david-yz-liu) [inline demo/src/App.tsx:29]: Add some error handling here as well, in the case that the `JSON.parse` fails.

File: demo/src/App.tsx [hunk 2]
```diff
@@ -20,13 +21,22 @@ const DEMO_OBJECTS = [
...
+    };
+
     return (
         <>
-            <section>
-                <h2>Input</h2>
-                <pre>{JSON.stringify(DEMO_OBJECTS)}</pre>
```
Comment (david-yz-liu) [inline demo/src/App.tsx:27]: good; you can delete the `DEMO_OBJECTS` constant as well

File: demo/src/MemoryModels.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    isClass: boolean;
+    name: string;
+    id: string | number | null;
+    value: any;
+    stack_frame: boolean;
+    show_indexes: boolean;
+    style?: any[] | null;
+}
+
+function MemoryModels(props) {
```
Comment (david-yz-liu) [inline demo/src/MemoryModels.tsx:14]: I'm not a big fan of this name. This component is really the user input component, right? Try picking a name that reflects that role

File: demo/src/MemoryModels.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    return (
+        <form onSubmit={props.onSubmit}>
+            <Card>
+                <CardContent>
+                    <Grid container spacing={2}>
+                        <Grid item xs={12}>
+                            <TextField
+                                id="multiline-textfield"
+                                label="Multi-line Text"
```
Comment (david-yz-liu) [inline demo/src/MemoryModels.tsx:27]: Looks like this is also used as a placeholder text. Something like `Enter memory model JSON here`

File: demo/src/MemoryModels.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        props.setFormData(event.target.value);
+    };
+
+    return (
+        <form onSubmit={props.onSubmit}>
+            <Card>
+                <CardContent>
+                    <Grid container spacing={2}>
+                        <Grid item xs={12}>
+                            <TextField
```
Comment (david-yz-liu) [inline demo/src/MemoryModels.tsx:25]: would be nice to have this use a monospace font

File: demo/src/MemoryModels.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                                style={{ width: "100%" }}
+                            />
+                        </Grid>
+                        <Grid item xs={12}>
+                            <Button
+                                type="submit"
+                                variant="contained"
+                                color="primary"
+                            >
+                                Submit
```
Comment (david-yz-liu) [inline demo/src/MemoryModels.tsx:44]: Change label to "Draw Diagram"

File: demo/src/SvgDisplay.tsx [hunk 7]
```diff
@@ -1,31 +1,33 @@
...
-    {
-        isClass: false,
-        name: "str",
-        id: 19,
-        value: "David is cool!",
-        style: ["highlight"],
-    },
-    { isClass: false, name: "int", id: 13, value: 7 },
-];
+// const DEMO_OBJECTS = [
```
Comment (david-yz-liu) [inline demo/src/SvgDisplay.tsx:4]: This is fine, you can delete now

File: package.json [hunk 8]
```diff
@@ -26,6 +26,7 @@
     "dependencies": {
         "@xmldom/xmldom": "^0.8.6",
         "deepmerge": "^4.3.1",
+        "react-error-boundary": "^4.0.12",
```
Comment (david-yz-liu) [inline package.json:29]: This should be in the `demo/package.json` file

File: demo/src/index.tsx [hunk 9]
```diff
@@ -3,10 +3,14 @@ import { createRoot } from "react-dom/client";
 import App from "./App";
 import "@picocss/pico";
 import "./css/styles";
+("use client");
```
Comment (david-yz-liu) [inline demo/src/index.tsx:6]: I don't think this is necessary in this case? We aren't doing any server-side rendering

File: demo/src/index.tsx [hunk 10]
```diff
@@ -3,10 +3,14 @@ import { createRoot } from "react-dom/client";
...
+import { ErrorBoundary } from "react-error-boundary";
 
 const root = createRoot(document.getElementById("app"));
 root.render(
     <StrictMode>
-        <App />
+        <ErrorBoundary fallback={<div>Something went wrong</div>}>
```
Comment (david-yz-liu) [inline demo/src/index.tsx:12]: So I was expecting this to wrap just the output (`SvgDisplay`), did you not get that to work?

File: demo/src/MemoryModels.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import React from "react";
+import { Button, Card, CardContent, TextField, Grid } from "@mui/material";
+
+interface MemoryModel {
```
Comment (david-yz-liu) [inline demo/src/MemoryModels.tsx:4]: I don't think this is being used anymore, so you can remove it.

File: demo/package.json [hunk 12]
```diff
@@ -33,6 +34,9 @@
         "webpack-cli": "^5.1.4"
     },
     "dependencies": {
+        "@emotion/react": "^11.11.3",
```
Comment (david-yz-liu) [inline demo/package.json:37]: It doesn't seem like you're actually using `@emotion`, in which case it's good to remove from this PR.

File: demo/src/index.tsx (hunk 10 above)
Comment (TheTallJerry) [reply demo/src/index.tsx:12]: Fixed, I moved the `Output` heading out of `SvgDisplay`

image

File: demo/package.json [hunk 13]
```diff
@@ -33,8 +34,11 @@
         "webpack-cli": "^5.1.4"
     },
     "dependencies": {
+        "@emotion/styled": "^11.11.0",
```
Comment (david-yz-liu) [inline demo/package.json:37]: I think you missed removing this

File: package-lock.json
Comment (david-yz-liu) [inline package-lock.json:1]: The changes in this file should be reverted (as you have reverted changes to the top-level )`package.json`

Comment (TheTallJerry) [reply package-lock.json:1]: Reverting them failed CI https://github.com/david-yz-liu/memory-models-rough/actions/runs/7878098310. Looks like `demo` is a sub-package to the parent `package-lock`, so `demo`'s dependencies might need to be preserved there.

**Review by david-yz-liu — APPROVED** [review]

**Pull request merged**